lib/archethic/mining/pending_transaction_validation.ex

```diff
@@ -64,12 +65,72 @@ defmodule Archethic.Mining.PendingTransactionValidation do
 
         {:error, "Invalid previous signature"}
 
+      {:error, :invalid_chain_type} ->
+        Logger.error("Chain type Error",
+          transaction_address: Base.encode16(address),
+          transaction_type: type
+        )
+
+        {:error, "Invalid previous Chain Type and Current Transaction Type"}
+
       {:error, reason} = e ->
         Logger.info(reason, transaction_address: Base.encode16(address), transaction_type: type)
         e
     end
   end
 
+  @spec validate_previous_transaction_type?(Archethic.TransactionChain.Transaction.t()) ::
+          :ok | {:error, :invalid_chain_type}
+  def validate_previous_transaction_type?(tx) do
+    val =
+      case Transaction.network_type?(tx.type) && tx.type not in [:code_proposal] do
+        false ->
+          #  not a network tx , no need to validate with
+          #  last tx
+          true
+
+        true ->
+          # when network tx , check with previous transaction
+          do_validate_previous_transaction_type?(tx)
+      end
+
+    if val == true, do: :ok, else: {:error, :invalid_chain_type}
+  end
+
+  @spec do_validate_previous_transaction_type?(Archethic.TransactionChain.Transaction.t()) ::
+          boolean
+  def do_validate_previous_transaction_type?(curr_txn = %Transaction{}) do
+    prev_address = Transaction.previous_address(curr_txn)
+
+    case TransactionChain.get_transaction(prev_address, [:address, :type]) do
+      {:error, :transaction_not_exists} ->
+        # previous txn do not exist, a genesis_txn
+        # TODO imp:
+        # impl to different b/w user generated netowrk_txn
+        true
+
+      {:error, :invalid_transaction} ->
+        # Last txn is
+        # TODO imp: impl handling of last KO txn
+        false
+
+      {:ok, last_txn = %Transaction{}} ->
+        match_last_tx_type?(curr_txn.type, last_txn.type)
+    end
+  end
+
+  @spec match_last_tx_type?(Transaction.transaction_type(), Transaction.transaction_type()) ::
+          boolean
+  def match_last_tx_type?(:oracle, :oracle_summary), do: true
```

Comment:
I don't think the cyclic mode is always true, for instance oracle can be followed by oracle and oracle summary but nothing after oracle_summary

Reply:
we check for current txn type and last txn type only, and we are relying on network bootstrap , for last txn type.